DMCompiler/DM/Expressions/Binary.cs

```diff
@@ -163,7 +163,13 @@ public LeftShift(Location location, DMExpression lhs, DMExpression rhs)
 
         public override void EmitPushValue(DMObject dmObject, DMProc proc) {
             LHS.EmitPushValue(dmObject, proc);
-            RHS.EmitPushValue(dmObject, proc);
+
+            if (RHS.TryAsConstant(out Constant constValue)) { // RHS could be a const-evaluable expr
+                constValue.EmitPushValue(dmObject, proc);
+            } else {
+                RHS.EmitPushValue(dmObject, proc);
+            }
```

Comment:
This is somewhat of a weird thing. right now, our constant system isn't set up where we can const-eval just the rhs of an expression (see: const << const, foo << const, foo << foo) it seems in the TryAsConstant tree - so we do it here in the EmitPushValue even though i think the proper behavior would be to TryConstant in `ProcessStatementExpression()`

Reply:
for the logical conclusion, see #885

Reply:
to be clear, this bit should still stay in the for the case of where the rhs can be constant but the lhs can't be - the TryAsConstant needs both to be true

see the case of `world << (5 + 5)`

Reply:
I don't think it should, #885 would lead both lhs and rhs to go through their own constant folding pass.

Reply:
I don't think it should, #885 would lead both lhs and rhs to go through their own constant folding pass.